rpc/gnmi/gnmi-history.md

```diff
@@ -0,0 +1,87 @@
+# gNMI History Extension
+
+**Contributors:** Justin Costa-Roberts, Aaron Beitch, Mike Fink
+
+**Date:** August 4, 2020
+
+**Version:** 0.1.0
+
+# 1. Purpose
+
+In certain applications it is necessary not only to retrieve current data from a
+data tree, but also to allow querying historical data. The gNMI History
+extension enables clients both to retrieve data at a specific time in the past,
+and to request all updates applied to a data tree between two specified times.
+
+# 2. Definition
+
+A `History` message is embedded in an Extension message in the
+`SubscribeRequest` proto.
+
+## 2.1 Proto
+
+```
+message TimeRange {
+  int64 start = 1;
+  int64 end = 2;
+}
+
+message History {
+  oneof request {
+    int64 time = 1;
+    TimeRange range = 2;
+  }
+}
+```
+
+## 2.2 SubscribeRequest handling
+
+The History extension may be applied only to `ONCE` and `STREAM` subscriptions.
+Requests for historical data can be either of two types: a snapshot at a
+specified time, or a set of updates between two times. The `History` message's
+request field is a `oneof` field whose type indicates which of these request
+types applies.
+
+In all cases, times are represented as int64s interpreted as nanoseconds since
+the Unix Epoch.
+
+### 2.2.1 Snapshot requests
+
+If the `time` field is set, the request is for a snapshot. The subscription mode
+for a time request must be `ONCE`; otherwise the server must return an
+`INVALID_ARGUMENT` error. The server must transmit the data indicated in the
+`SubscribeRequest` as it existed on the server at the specified time. If the
+specified time indicates a future time at the point the server receives the
+request, the server may either return an `UNIMPLEMENTED` error or choose to wait
+until the specified time has been reached.
+
+### 2.2.2 Stream requests
```

Comment:
Does the stream include an initial snapshot? or only updates received between the date ranges?

Reply:
I tried to address this below, with "Whether the server should transmit the specified data tree as it existed at the start time is determined as usual by the `updates_only` field on the `SubscriptionList` message."

Maybe a clearer way to say this is "As usual, the server should also transmit the specified data tree as it existed at the subscription start time unless the `updates_only` field on the `SubscriptionList` message is set." What do you think?

Reply:
This sounds good to me.